README.md
# Budget Shary

[CHANGELOG](/CHANGELOG.md)<br/>
**[LIVE DEMO](https://shary.wailorman.ru/)**<br/>
[BUILDS](https://32225.selcdn.ru/budget-shary/builds/)

## What is this?
Goal of this application is to equalize _expenses_ of all persons.

For example, two guys (Mike and Jimmy) want to drink a beer at home. Mike
went to a shop and buy two bottles of beer and spent 60 dollars (it's very expensive beer).

The $60 price is High for Mike. He wants to split it with Jimmy to not be left without money.

#### How?

So.
Both guys have the same _shares_ in this _budget_ -- 50%.
Mike have $60 expenses. Jimmy -- $0 expenses.
That is, _monetary shares_ of each guy = $30 _(totalExpenses * share)_

We need to equalize own expenses and monetary shares for each person.

By this example, funds of
```
Mike = 30 - 60 = -30
Jimmy = 30 - 0 = +30
```

Funds calculated by this formula:
```
Funds = shareInMonetary - ownExpenses [ + incomeTransactionsTotal - outcomeTransactionsTotal ]
```

To make funds of each persons equals 0, we can follow two ways:
  1. Reduce monetary share of Jimmy to 0 and increase for Mike to 60.<br/>
     But this way Mike will become drunk.
  2. Reduce expenses of Mike to 30 and increase for Jimmy to 30.

I think, Jimmy will like second way. So how to change expenses?<br/>
Easy! Jimmy should give $30 to Mike. In other language, we
need to create a transaction from Jimmy to Mike with 30 dollars total.

After transaction has been proceeded, let's recalculate funds for beer-guys.
```
Mike = 30 - 60 + 30 - 0 = 0
Jimmy = 30 - 0 + 0 - 30 = 0
```

Funds are equalized and now guys are happy.


## How to run?

##### **`npm start`**
Developing mode. Starting webpack-dev-server and watch on changes in all files (including sources & tests)
* [unit-browser.build](http://localhost:8080/webpack-dev-server/unit-browser.build) 
-- Live-reloading unit tests in browser
* [integration-dev.build](http://localhost:8080/webpack-dev-server/integration-dev.build) 
-- Live-reloading integration tests in browser
* [index.html](http://localhost:8080/webpack-dev-server/dist/index.html) 
-- Live-reloading main app page

Keeps tests files prebuilded

##### **`npm test`**
Runs all tests

##### **`[NODE_ENV=production] npm run build`**
Runs webpack and compiles all assets (sources & tests)

##### **`npm test:unit`**
Run unit tests (prebuild required)

##### **`npm run test:integration`**
##### **`npm run test:integration:dev`**
Runs integration tests in Karma and PhantomJS. Use `:dev` suffix to run Karma in watch mode

##### **`npm run storybook`**
Runs [Storybook](https://github.com/kadirahq/react-storybook) on 9001 port

CHANGELOG.md
# 2.3.0-beta
* **Materialize interface** and other restylings
* Add PWA manifest.json
* (bug-fix) No more errors with creating new persons/products with existing ids ([3df9031](https://github.com/wailorman/budget-shary/commit/3df9031c8416254f0adc483c7c6adead235e17f9),
                                                                                 [3e0dd64](https://github.com/wailorman/budget-shary/commit/3e0dd64d9d17eabd0e21aab6b265c30485c95b70),
                                                                                 [9de30e5](https://github.com/wailorman/budget-shary/commit/9de30e57fecb14c4d3fd4b6aa3b5e10411f72694))
* (dev) Auto-deploy all travis builds to Selectel (see [77d3df2](https://github.com/wailorman/budget-shary/commit/77d3df25fa20978a7a23894abcc17101f1e03f72))
* (dev) Optimize rendering (More containers, better _shouldComponentUpdate()_)


# 2.2.0-beta
* Bottom margin after participating buttons on phones ([8f33731](https://github.com/wailorman/budget-shary/commit/8f33731))
* **Multiple budgets**.<br/>
  Now you should not to erase previous budget to create a new one. You can just
  create new budget, keeping an old one. 
* (bug-fix) Calculating can't be proceeded if exists empty person with empty product(s) ([a98f518](https://github.com/wailorman/budget-shary/commit/a98f518))
* (bug-fix) productParticipating spamming into console ([7de8d33](https://github.com/wailorman/budget-shary/commit/7de8d33))
* Write little description about app interchanging algorithm ([bd6b6dc](https://github.com/wailorman/budget-shary/commit/bd6b6dc))
* (dev) Write good JSDocs for core functions ([15e3c06](https://github.com/wailorman/budget-shary/commit/15e3c06), 
                                              [dd66775](https://github.com/wailorman/budget-shary/commit/dd66775))

# 2.1.0-beta
* **Participating feature**.<br/> 
  Now you can just click on persons which are have share in particular product and 
  share will be calculated automatically
* (dev) Better, more predictable and simple algorithm of validation
* (dev) Interchange proceeding moved to own middleware
* (dev) Remove api code from repository. I'll create independent repository for api
* (dev) Reorganize npm scripts (read the [new README.md](https://github.com/wailorman/budget-shary/blob/110682e45078b4af56d58e06b7d83fa7dc832cad/README.md))


src/state-stub.js
import shortid from 'shortid';
export const STUB_BUDGET_ID = 'stub';

export const generateBudget = ()=> {
    return {
        budget: {
            id: `_b_${shortid.generate()}`,
            name: ''
        },
        persons: {},
        products: {},
        productParticipating: {},
        transactions: [],
        errors: {
            products: {},
            persons: {},
            common: {}
        }
    };
};

export const stateStub = {

    "budget": {
        "id": STUB_BUDGET_ID,
        "name": "Example budget"
    },
    "products": {
        "_product_ryrGguKzg": {
            "id": "_product_ryrGguKzg",
            "name": "Cola",
            "price": "150",
            "ownerId": "_person_S1tZg_KMx"
        },
        "_product_SJtzedKfx": {
            "id": "_product_SJtzedKfx",
            "name": "Chips",
            "price": "50",
            "ownerId": "_person_S1tZg_KMx"
        },
        "_product_rkkXxdKzx": {
            "id": "_product_rkkXxdKzx",
            "name": "Cake",
            "price": "300",
            "ownerId": "_person_S1tZg_KMx"
        },
        "_product_SJSQluYGg": {
            "id": "_product_SJSQluYGg",
            "name": "Flowers",
            "price": "300",
            "ownerId": "_person_ryCZe_tMx"
        },
        "_product_BJYQgOKfg": {
            "id": "_product_BJYQgOKfg",
            "name": "Beer",
            "price": "200",
            "ownerId": "_person_H1Wfe_FMl"
        },
        "_product_Hyk4gdYfl": {
            "id": "_product_Hyk4gdYfl",
            "name": "Nuts",
            "price": "40",
            "ownerId": "_person_H1Wfe_FMl"
        },
        "_product_HyC4eutzx": {
            "id": "_product_HyC4eutzx",
            "name": "Candles",
            "price": "30",
            "ownerId": "_person_ryCZe_tMx"
        }
    },
    "persons": {
        "_person_S1tZg_KMx": {
            "id": "_person_S1tZg_KMx",
            "name": "Mike",
            "share": 28.504672897196258
        },
        "_person_ryCZe_tMx": {
            "id": "_person_ryCZe_tMx",
            "name": "Alice",
            "share": 28.971962616822427
        },
        "_person_H1Wfe_FMl": {
            "id": "_person_H1Wfe_FMl",
            "name": "Jimmy",
            "share": 42.523364485981304
        }
    },
    "productParticipating": {
        "_product_ryrGguKzg": {
            "_person_S1tZg_KMx": true,
            "_person_H1Wfe_FMl": true,
            "_person_ryCZe_tMx": true
        },
        "_product_SJtzedKfx": {
            "_person_S1tZg_KMx": true,
            "_person_H1Wfe_FMl": true
        },
        "_product_rkkXxdKzx": {
            "_person_ryCZe_tMx": true,
            "_person_H1Wfe_FMl": true
        },
        "_product_SJSQluYGg": {
            "_person_S1tZg_KMx": true,
            "_person_H1Wfe_FMl": true,
            "_person_ryCZe_tMx": true
        },
        "_product_HyC4eutzx": {
            "_person_S1tZg_KMx": true,
            "_person_ryCZe_tMx": true,
            "_person_H1Wfe_FMl": true
        },
        "_product_BJYQgOKfg": {
            "_person_S1tZg_KMx": true,
            "_person_H1Wfe_FMl": true
        },
        "_product_Hyk4gdYfl": {
            "_person_S1tZg_KMx": true,
            "_person_H1Wfe_FMl": true
        }
    },
    "transactions": [
        {
            "from": "Jimmy",
            "to": "Mike",
            "total": 200.4
        },
        {
            "from": "Jimmy",
            "to": "Alice",
            "total": 9
        }
    ]

};